feat: components are now easier to identify

steven-murray opened the pull request:
This updates the components such that a base component tracks its sub-models, but *also* gets tracked, so that all models and components can be found more easily. It also adds some nice functions to list these models, and updates appropriate parts of the `Simulator` class to use this new functionality.

r-pascua commented on hera_sim/components.py:
```diff
@@ -1,14 +1,18 @@
 import re
-from abc import ABCMeta, abstractmethod
-
+from abc import abstractmethod
```
Has there been a change with `abstractmethod` that I'm unaware of? At the time of writing, a class was required to have `ABCMeta` as its metaclass in order to actually use `abstractmethod` as intended. I'm not entirely convinced that the `abstractmethod` requirement is being enforced in subclasses. I tested the component discoverability with a custom class as follows:
```
from hera_sim import component

@component
class Test:
    pass

class Model(Test):
    pass
```
and this worked without a hitch, which suggests that the `abstractmethod` decorator isn't actually doing anything. (If my understanding is correct, then decorating the `__call__` method with `abstractmethod` makes it so that any subclass *must* define its own `__call__` method. If I'm wrong, then please correct me on that point.) On another note, though, the component was discovered via `get_all_components`.

steven-murray replied:
Ah. Yeah that's my mistake.

r-pascua reviewed (changes requested):
There is a lot of great stuff in here, and it looks like the tracking effect is still achieved with less code, so thank you for your work! I have a few concerns that I would like addressed before approving this PR--please see the comments for details. Going over this PR also made me realize there are some things I'd like to discuss regarding designation of components (e.g. should an `Array` really be considered a component?) and construction of the `_available_components` dictionary (in particular, capitalization of the keys). I think the first point is outside the scope of this PR, but the second one might not be. Curious to hear your thoughts.

Thanks again--these updates are great!

steven-murray commented on hera_sim/simulate.py:
```diff
@@ -754,64 +761,29 @@ def _get_model_parameters(model):
+            except KeyError:
+                raise ValueError(
+                    f"The model '{component}' does not exist. The following models are"
+                    f"available: {get_all_components(with_aliases=True)}."
```
Ah, yeah. I think I can write a little function that makes this output nicer. Which aliases do we want to list? I could do something like:

```
component_name:
   model_name / model_alias / other_alias
   other_model / its_alias
other_component:
  ...
```

steven-murray replied:
I did this in dce7100 like I stated in the comment above. I get this: 

```
Array:
  LinearArray | lineararray | linear_array
  HexArray | hexarray | hex_array
Foreground:
  DiffuseForeground | diffuseforeground | diffuse_foreground
  PointSourceForeground | pointsourceforeground | point_source_foreground | pntsrc_foreground
Noise:
  ThermalNoise | thermalnoise | thermal_noise
RFI:
  Stations | stations | rfi_stations
  Impulse | impulse | rfi_impulse
  Scatter | scatter | rfi_scatter
  DTV | dtv | rfi_dtv
Gain:
  Bandpass | bandpass | gains | bandpass_gain
  Reflections | reflections | reflection_gains | sigchain_reflections
Crosstalk:
  CrossCouplingCrosstalk | crosscouplingcrosstalk | cross_coupling_crosstalk | cross_coupling_xtalk
  CrossCouplingSpectrum | crosscouplingspectrum | cross_coupling_spectrum | xtalk_spectrum
  WhiteNoiseCrosstalk | whitenoisecrosstalk | white_noise_crosstalk | whitenoise_xtalk | white_noise_xtalk
EoR:
  NoiselikeEoR | noiselikeeor | noiselike_eo_r | noiselike_eor
```

Still pending taking out other aliases. Does this look good to you?

r-pascua replied:
This is formatted very nicely! I definitely like how the output looks.

steven-murray commented on hera_sim/components.py:
```diff
@@ -19,152 +23,189 @@ class SimulationComponent(metaclass=ABCMeta):
+def component(cls):
+    """Decorator to create a new specific Component that tracks its models."""
+    cls._models = {}
+    cls = new_class(
+        cls.__name__,
+        (SimulationComponent,),
+        {"is_abstract": True},
+        lambda ns: ns.update(dict(cls.__dict__)),
+    )
```
Yeah, I'd never used this function before. It's certainly arcane.

steven-murray replied:
Added a whole comment above it, and the keywords. Let me know if that looks good.

r-pascua replied:
Thank you! It is very clear what is happening now.

steven-murray commented on hera_sim/components.py:
```diff
@@ -19,152 +23,189 @@ class SimulationComponent(metaclass=ABCMeta):
+
+    @classmethod
+    def get_aliases(cls):
+        return (
+            cls.__name__,
+            cls.__name__.lower(),
+            inflection.underscore(cls.__name__),
+        ) + cls._alias
```
I'm completely fine with just keeping the original class names and the lower case. I agree that an abundance of names makes things tricky. I guess we need to keep the aliases at this stage for backwards compat? should I just remove the underscore version then?

steven-murray replied:
I have _not_ done this in dce7100. I'm still not sure exactly which aliases we'd like to keep around. It will be easy once we decide that.

r-pascua replied:
I think keeping the original class names and the lowercase, as well as the backwards compat names, is a good choice. It gives a decent amount of options without making too much clutter. We could even *only* put in the lowercase names and the backwards compat names, and have the `Simulator` only look for lowercase matches when adding/retrieving a component using a string.

steven-murray replied:
OK, excellent. I think I've done this now.

steven-murray commented on hera_sim/components.py:
```diff
@@ -1,14 +1,18 @@
 """
-
-import functools
+from __future__ import annotations
```
This is not a Py2 thing -- it's something that I think was put in 3.8 or 3.9 and backported to 3.7+. It does make 3.6 break, but I'd argue that no-ones still using that. 

What this in particular does is means you can use type annotations referring to a given class name _inside_ that class. Without it, it would say that the class is not yet defined. I can just take out these annotations, but I don't want to :-)

r-pascua replied:
If you think 3.6 is no longer being used, then I think it's okay to leave this in--it sounds like a nice feature. Though we'll need to drop 3.6 testing.

steven-murray commented:
What are your thoughts on the capitilization of keys in `_available_components`? Do you think they should be capitalized? At the moment it's just putting things in exactly as the class is defined. I actually typically like putting them in all `.lower()` so that user input can be converted to `.lower()` and matched.

steven-murray commented on hera_sim/simulate.py:
```diff
@@ -754,64 +761,29 @@ def _get_model_parameters(model):
+                return get_model(component), True
+            except KeyError:
+                raise ValueError(
+                    f"The model '{component}' does not exist. The following models are"
```
Addressed in dce7100

steven-murray commented on hera_sim/simulate.py:
```diff
@@ -754,64 +761,29 @@ def _get_model_parameters(model):
+            raise ValueError(
+                "The input type for the component was not understood. "
+                "Must be str, Type[SimulationComponent]"
+                "or an instance of a SimulationComponent"
```
I addressed this in dce7100. Added a print out of the parameters as well as a slightly nicer description of the inputs.

steven-murray commented on hera_sim/simulate.py:
```diff
@@ -754,64 +761,29 @@ def _get_model_parameters(model):
+        else:
+            raise ValueError(
+                "The input type for the component was not understood. "
+                "Must be str, Type[SimulationComponent]"
```
Yup, thanks. Done in dce7100.

steven-murray commented on hera_sim/simulate.py:
```diff
@@ -754,64 +761,29 @@ def _get_model_parameters(model):
-                    "``hera_sim`` class or an alias thereof."
+    def _get_component(
+        component: [str, Type[SimulationComponent], SimulationComponent]
+    ) -> Union[SimulationComponent, Type[SimulationComponent]]:
```
You're right. Done in dce7100.

steven-murray commented on hera_sim/components.py:
```diff
@@ -19,152 +23,189 @@ class SimulationComponent(metaclass=ABCMeta):
+        if with_aliases:
+            return deepcopy(cls._models)
+        else:
+            return {c.__name__: c for c in set(cls._models.values())}
```
Done, thanks.

r-pascua commented on hera_sim/simulate.py:
```diff
@@ -754,64 +761,29 @@ def _get_model_parameters(model):
+            except KeyError:
+                raise ValueError(
+                    f"The model '{component}' does not exist. The following models are "
+                    f"available: \n{print_all_components()}."
```
Using `print_all_components()` here returns `None`, and so the output of this error message isn't what's expected. We've got (at least) two options: make `print_all_components` return the string literal, or turn `print_all_components` into a thin wrapper for a function `list_all_components` that does essentially what `print_all_components` currently does, but returns the string literal rather than printing it.

steven-murray replied:
Good catch. I chose to rename the function to list_all_components and update it to return the string. I think printing inside the function is kinda silly.

r-pascua commented on hera_sim/simulate.py:
```diff
@@ -754,64 +761,29 @@ def _get_model_parameters(model):
+            return component, False
+        else:
+            raise ValueError(
+                "The input type for the component was not understood. "
+                "Must be a string, or a class/instance of type 'SimulationComponent'. "
+                f"Available component models are:\n{print_all_components()}"
```
This is much more informative than before, thank you! The last part of the error message isn't what's expected--see previous comment.

steven-murray replied:
Fixed as above.

r-pascua commented:
A lot of the changes I requested have been made, and it's looking much better now. Thanks for the quick turnaround! There are a few more changes I'd like to see before approving though: 
* There is a bug with `get_models` and `get_all_models` that I would like to see addressed; 
* There is a bug with the error messages in `Simulator._get_component` relating to `print_all_components` returning `None`; 
* We need to remove Py3.6 testing in order to keep the `__future__.annotations` import; 
* I'm still leaning toward trimming the number of aliases that are tracked.
We also need to increase the diff coverage, but I can do this once the requested changes are made. Thanks you for the great work!

steven-murray commented on hera_sim/components.py:
```diff
@@ -19,152 +23,189 @@ class SimulationComponent(metaclass=ABCMeta):
+    return get_all_components(with_aliases)[cmp.__name__]
+
+
+def get_all_models(with_aliases: bool = False) -> Dict[str, SimulationComponent]:
+    all = get_all_components(with_aliases)
+    out = {}
+    for k, v in all.items():
+        out.update(v)
+    return out
```
I updated the variable names as suggested, and also added some basic docstrings and comments to make it more obvious what's happening.

steven-murray commented:
Thanks for the review @r-pascua. I think I've addressed all your points. For the aliases, it now only tracks the lower-case class name and any explicitly-set aliases. When looking for the model, the input string is lower-cased so it is essentially case-insensitive.

r-pascua reviewed (approved):
It looks like everything I've asked for has been done, and the error message spit out when trying to add a nonexistent component is very nicely formatted. Thanks for the great work!
